cmd/holo-files/internal/common/filebuffer.go

```diff
@@ -21,79 +22,82 @@
 package common
 
 import (
-	"bytes"
 	"errors"
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"syscall"
 )
 
-//FileBuffer represents the contents of a file. It is used in holo.Apply() as
+// FileBuffer errors
+var (
+	ErrNotManageable = errors.New("not a manageable file")
+	ErrExist         = errors.New("target exists and is not a manageable file")
+)
+
+//FileBuffer represents a file, loaded into memory. It is used in holo.Apply() as
 //an intermediary product of application steps.
 type FileBuffer struct {
-	//set only for regular files
-	Contents []byte
-	//set only for symlinks
-	SymlinkTarget string
-	//used by ResolveSymlink (see doc over there)
-	BasePath string
+	Path     string
+	Mode     os.FileMode
+	UID      int
+	GID      int
+	Contents string
+
+	Manageable bool
 }
 
 //NewFileBuffer creates a FileBuffer object by reading the manageable file at
-//the given path. The basePath is stored in the FileBuffer for use in
-//holo.FileBuffer.ResolveSymlink().
-func NewFileBuffer(path string, basePath string) (*FileBuffer, error) {
-	info, err := os.Lstat(path)
+//the given path.
+func NewFileBuffer(path string) (FileBuffer, error) {
+	return newFileBuffer(path, false)
+}
+
+func newFileBuffer(path string, follow bool) (fb FileBuffer, err error) {
+	fb.Path = path
+
+	var info os.FileInfo
+	if follow {
+		info, err = os.Stat(path)
+	} else {
+		info, err = os.Lstat(path)
+	}
 	if err != nil {
-		return nil, err
+		return
 	}
 
-	//a manageable file is either a symlink...
-	if IsFileInfoASymbolicLink(info) {
-		target, err := os.Readlink(path)
+	stat := info.Sys().(*syscall.Stat_t) // UGLY
+
+	fb.Mode = info.Mode()
+	fb.UID = int(stat.Uid)
+	fb.GID = int(stat.Gid)
+
+	if fb.Mode&os.ModeSymlink != 0 {
```

Comment:
I think this will also fire for regular files. In a POSIX-compatible `struct stat`, the file-mode bits are `010000` for regular files and `012000` for symlinks, and `010000 & 012000 != 0`. I suggest you use the same test as in `IsFileInfoASymlink`. (What confuses me is that if this really is a problem, it should've been caught by the tests many times.)